### PR TITLE
Implement Smart Pointer Creation With Default Initialization

### LINKED ISSUE #48: P1020R1 Smart Pointer Creation With Default Initialization
[P1020R1](https://wg21.link/P1020R1) Smart Pointer Creation With Default Initialization

[P1973R1](https://wg21.link/P1973R1) Renaming `default_init` To `for_overwrite`

Final feature-test macro in WG21-N4861:

`#define __cpp_lib_smart_ptr_for_overwrite 202002L`

### PR DESCRIPTION
**Pull request by AdamBucior**
Resolves #48. Based on #778.

### CODE REVIEW

File: stl/inc/memory
Comment (miscco) [inline stl/inc/memory:2960]: I would prefere if you swap the blocks. It is always easier to follow `meow` rather than `!meow`

File: stl/inc/memory
Comment (miscco) [inline stl/inc/memory:3102]: Same here and throughout

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory]: Done

**Comment by StephanTLavavej** [comment]
I pushed a commit (deleting an extra newline) with the `--author=` option, so when we merge this PR, it will be displayed as co-authored by @Weheineman. Thanks for picking up this work!

**Comment by StephanTLavavej** [comment]
Status update: Planning to review this on Thursday, Oct 8. Apologies for the delay!

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Thanks! This looks good, my main concern is with a potential corner case in overload resolution that should be avoidable with different metaprogramming. Please let me know if you'd like me to explain further, or I can revise the PR if you want - your preference.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (TylerMSFT) [inline tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:75]: Could you turn the magic number into a const, with a readable name? It doesn't change the functionality - just makes it a little more readable.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (TylerMSFT) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:122]: Could you turn the magic number 0xEE into a const, with a readable name? It doesn't change the functionality - just makes it a little more readable.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (TylerMSFT) [inline tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:38]: Could you turn the magic number 0xEE into a const, with a readable name? It doesn't change the functionality - just makes it a little more readable.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (TylerMSFT) [inline tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:58]: Could you turn the magic number 0xEE into a const, with a readable name? It doesn't change the functionality - just makes it a little more readable.

File: stl/inc/memory
Comment (StephanTLavavej) [reply stl/inc/memory:2195]: @cdacamar noted that we can deal with the variadic `_Types` without introducing a wrapper class, by using fold expressions. That is, `(is_same_v<_Types, _For_overwrite_tag> && ...)` (I believe that's the syntax) combined with `sizeof...(_Types) == 1` should be sufficient.

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:153]: Done

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:85]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:141]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Done

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory]: Done

File: tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp
Comment (TylerMSFT) [inline tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp:19]: The meaning of 'defaultValue` is very broad. Perhaps: initializedMemoryValue?

Comment (AdamBucior) [reply tests/std/tests/P1020R1_smart_pointer_for_overwrite/test.cpp]: Renamed to `initializedValue`

**Comment by AdamBucior** [comment]
It's ready for another review.

**Comment by TylerMSFT** [comment]
@AdamBucior, thank you for addressing my feedback. I'm good with your changes. I defer to Stephan for the remaining review.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
This looks great - thanks for the thorough and highly detailed work. (I noticed and appreciated the improvements like using `remove_cvref_t` on perfectly forwarded `_Types` to handle const lvalues, and unifying additional repetitive code. Combining 3 overloads into 1, for both the `make` and `allocate` families, is awesome.)

I pushed a merge with `master` and a small set of superficial cleanups, after verifying that the test is still passing:

* When granting friendship to `_Make_shared_unbounded_array` and `_Allocate_shared_unbounded_array`, drop `const` on the value parameter `size_t _Count`. (We conventionally avoid top-level `const` on value parameters in function declarations, since it has no meaning to callers and is immediately ignored by the language. We are so picky about `const`! :joy_cat:)
* Your `_Make_shared_unbounded_array`/`_Allocate_shared_unbounded_array` refactoring has a bonus: we no longer need to grant friendship to several `make_shared` etc. overloads.
* When functions take `const _Arg&`, we don't need to apply `remove_cvref_t` to `_Arg` (unless we were worried about `_Arg` being passed as an explicit template argument, instead of being deduced, which is not a concern here).
* Remove a leftover comment after the `allocate_shared` refactoring.
* Use `is_bounded_array_v` in `make_unique_for_overwrite` for consistency.

I'll find a second maintainer to review and then we can merge this! :rocket:

File: stl/inc/memory
Comment (miscco) [inline stl/inc/memory:2386]: Should we simplify this to
```suggestion
            _Uninitialized_default_construct_n(_Out, _Size); 
```

File: stl/inc/xutility
Comment (miscco) [inline stl/inc/xutility:132]: I think we should use `_Voidify_iter`
```suggestion
    ::new (_Voidify_iter(_STD addressof(_Obj))) _Ty;
```

File: stl/inc/xutility
Comment (miscco) [inline stl/inc/xutility:131]: I really do not like the name, maybe `_Default_construct_in_place`?

If you do change it, would you mind using it for the `uninitialized_default_construct{_n}` family

File: stl/inc/memory
Comment (miscco) [inline stl/inc/memory:2374]: To be absolutely sure, even for the `meow_overwrite` we do not want to default initialize here?

Comment (AdamBucior) [reply stl/inc/memory:2374]: That's the whole point of `*_for_overwrite`

File: stl/inc/memory
Comment (AdamBucior) [reply stl/inc/memory:2386]: There is no such function as `_Uninitialized_default_construct_n` and we can't use `uninitialized_default_construct_n` because it doesn't destroy elements in the reverse order.

File: stl/inc/xutility
Comment (AdamBucior) [reply stl/inc/xutility]: Done

File: stl/inc/xutility
Comment (AdamBucior) [reply stl/inc/xutility]: Done

File: stl/inc/memory
Comment (miscco) [reply stl/inc/memory:2381]: Hm there is the ranges version (where I found a unneeded specialization that I forgot to remove so thanks ;)) 

Should we also destroy in backwards order there?

Comment (AdamBucior) [reply stl/inc/memory:2381]: `ranges::uninitialized_default_construct_n` needs to support forward iterators and destroying in reverse order with them would be hard.

File: stl/inc/memory
Comment (CaseyCarter) [reply stl/inc/memory:2369]: For better or worse, there are many places in the STL where we assume that we can omit default-initialization of objects that are trivially default constructible. I'm not sure the resulting behavior is portably defined, but it has been working as intended on our implementation. We may need to make changes if and when compilers complain about this trick in `constexpr`, but that hasn't happened yet and in any case would need for more widespread changes than just this PR.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/yvals_core.h
Comment (StephanTLavavej) [reply stl/inc/yvals_core.h]: Thanks for catching this!

**Review by CaseyCarter — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks - the latest changes look good. I was concerned about debug codegen impact from increased use of `_Voidify_iter` (as we would `addressof` twice in several cases), but it does centralize the obnoxious casts. So I believe a reasonable compromise is to add a simple throughput optimization, avoiding deref-and-`addressof` when the input is already a raw pointer, and I went ahead and pushed that. For precedent, we do the same thing in `_Unfancy` although without `if constexpr`:
https://github.com/microsoft/STL/blob/e7d5113376763268075d87dbad662a7cf4763a10/stl/inc/xstddef#L281-L290

**Comment by StephanTLavavej** [comment]
Thanks again for implementing this feature, @AdamBucior and @Weheineman! This will help users achieve maximum performance with Standard smart pointers. :rocket:

**Pull request merged**
